go/src/meguca/geoip/db.go

```diff
@@ -59,3 +83,137 @@ func LookUp(ip string) (iso string) {
 
 	return
 }
+
+// check if the GeoLite2 country DB exists, and download if it doesn't
+func check() error {
+	// Get the MD5 hash from the DB and the new hash from upstream
+	oldHash, err := db.GetGeoMD5()
+
+	if err != nil {
+		return err
+	}
+
+	respMD5, err := http.Get("https://geolite.maxmind.com/download/geoip/database/GeoLite2-Country.tar.gz.md5")
+
+	if err != nil {
+		return err
+	}
+
+	defer respMD5.Body.Close()
+	md5Bytes, err := ioutil.ReadAll(respMD5.Body)
+
+	if err != nil {
+		return err
+	}
+
+	newHash := string(md5Bytes)
+	diff := !(oldHash == newHash)
+
+	// Ensure the HTTP response is a MD5 hash when converted
+	if len(newHash) != 32 {
+		return errors.New("response is not an MD5 hash")
+	}
+	
+	// Check if the GeoLite DB exists
+	_, err = os.Stat("GeoLite2-Country.mmdb")
+
+	if err != nil || diff {
+		if os.IsNotExist(err) || diff {
+			// Create the temporary archive and directory
+			tmpDir, err := ioutil.TempDir("", "tmp-")
+
+			if err != nil {
+				return err
+			}
+
+			defer os.RemoveAll(tmpDir)
+			tmp, err := ioutil.TempFile(tmpDir, "tmp-")
+
+			if err != nil {
+				return err
+			}
+			
+			// Get the archive itself
+			resp, err := http.Get("https://geolite.maxmind.com/download/geoip/database/GeoLite2-Country.tar.gz")
+
+			if err != nil {
+				return err
+			}
+
+			defer resp.Body.Close()
+			// Check if the tar.gz MD5 checksum matches the MD5 checksum we just downloaded
+			bodyBytes, err := ioutil.ReadAll(resp.Body)
+
+			if err != nil {
+				return err
+			} else if fmt.Sprintf("%x", md5.Sum(bodyBytes)) != newHash {
+				return errors.New("GeoLite DB MD5 checksums do not match")
+			}
+
+			// Write the response data to the temporary tar.gz and check the archive
+			_, err = tmp.Write(bodyBytes)
+
+			if err != nil {
+				return err
+			}
+
+			err = checkArchive(tmpDir, tmp, newHash)
+
+			if err != nil {
+				return err
+			}
+
```

Comment:
Why not load new database now?

Reply:
I feel it makes far more sense to keep the actual loading of the database in the `Load` function.

Reply:
Yes, but then you need to restart the server to load the DB otherwise.

Reply:
In this case, should we just move the entire `check` function into `Load`?

Reply:
I just don't see the point, `check` is a local function, while `Load` is an exported function, `Load` uses `check`. Are you saying you want to export `check` so you can check the status of the DB on-disk/md5 difference while meguca is running?
Should we have an exported `Reload` function for this instead?